### ISSUE TITLE
suggestion: change output device per application

### ISSUE THREAD

**Issue by erisdev** [kyleneideck/BackgroundMusic#79]
for instance, usually when my computer is plugged into a tv i'm either watching a movie or playing a game. i want to hear the game/movie/whatever audio on my tv but not other applications' sounds. and yet, i don't want them totally muted either...

**Comment by kyleneideck** [comment]
Good idea. It bugs me when programs don't have a way to change their audio device too, but it hadn't occurred to me that we could do this.

I imagine we'd implement it something like this:

* In BGMApp, the user changes the output device, say to a HDMI device that's connected to a TV, for an app, say MPlayer.
* BGMApp tells BGMDriver.
* BGMDriver creates a new input stream coming from BGMDevice, if it doesn't already have one for the HDMI output device, and redirects MPlayer's audio to that stream.
* BGMApp gets MPlayer's audio from that stream and plays it on the HDMI device.

By the way, in case you're not already aware of this, VLC lets you set the audio output device it uses. Though it wouldn't help with the games, of course.

**Comment by erisdev** [comment]
ooh, good tip with vlc. i'm typically streaming stuff in chrome but i suspect i could open most of the things i stream thru vlc too, either directly or with livestreamer. :+1:

**Comment by JohnTheodore** [comment]
This feature would be awesome. I only have a laptop, it would be nice to send music on chrome (google music) to one audio device (the larger room speakers over hdmi), and steam games (to another audio device, just the bluetooth speaker near the couch).

**Comment by ronggong** [comment]
@kyleneideck I am curious how would you do to direct audio of a specific app, say chrome, to a stream?

**Comment by kyleneideck** [comment]
@ronggong CoreAudio HAL plug-ins like BGMDriver can ask to receive the audio from each CoreAudio client separately. (See `kAudioServerPlugInIOOperationProcessOutput` in AudioServerPlugIn.h.) Generally, there will be one client for each process that plays audio.

Chrome has multiple processes, so it usually has multiple CoreAudio clients, but BGMDriver would be able to tell which clients belong to Chrome. Though, it can't always do that for every application.

When BGMDriver receives the audio for Chrome's clients, it could copy/mix it into a buffer that it uses just for Chrome's audio. So BGMDriver could give the Background Music device (BGMDevice) a new output stream and when CoreAudio (specifically, `coreaudiod`, sometimes called "the host" because it hosts the HAL plug-ins) asks BGMDriver for audio data for that stream, BGMDriver would give it the data from the buffer that has the audio from Chrome.
